### PR TITLE
A new tickets system to keep Cubes loaded

### PR DESCRIPTION
**Pull request by xcube16**
(I did not review it vary well after rebasing, so there may be bugs)

### CODE REVIEW

**Comment by Barteks2x** [comment]
Build fails

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java [hunk 1]
```diff
@@ -67,18 +67,10 @@ private boolean canUpdateEntity(World _this, int startBlockX, int oldStartBlockY
...
-		//otherwise entities would no longer be updated after falling out of the world
-		BlockPos entityPos = new BlockPos(updateEntity_entityPosX, updateEntity_entityPosY, updateEntity_entityPosZ);
-		if (!isValid(entityPos)) {
-			return true;
-		}
-		return isAreaLoaded(startBlockX, startBlockY, startBlockZ, endBlockX, endBlockY, endBlockZ, allowEmpty);
+		return this.isRemote() ||
+				this.getCubeFromBlockCoords(
+						new BlockPos(updateEntity_entityPosX, updateEntity_entityPosY, updateEntity_entityPosZ))
+						.getTickets().shouldTick();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java]: This is broken and less readable now.

And it loads cubes. It should NEVER load cubes.

First: it doesn't tick entities when they are outside of world border - once you fall out of the world you are stuck there forever.

Second - leave the 32 block thing. Entities can access nearby blocks, sometimes a lot of them like the ender dragon. They need some buffer radius.

File: src/main/java/cubicchunks/server/ChunkGc.java [hunk 2]
```diff
@@ -55,13 +53,6 @@ public void tick() {
...
 	private void chunkGc() {
-		for (Chunk chunk : cubeCache.getLoadedChunks()) {
-			Column column = (Column) chunk;
-			Collection<Cube> cubes = column.getLoadedCubes();
-			cubes.stream().filter(c -> !cubeMap.contains(c.getCoords())).forEach(cubeCache::unloadCube);
-			if(!cubeMap.contains(column.getX(), column.getZ())) {
-				cubeCache.unloadColumn(column);
-			}
-		}
+		cubeCache.takeOutGarbage();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/ChunkGc.java]: ...

The whole point of ChunkGc being separate class was that the unloading logic was in that class.

Now you moved it to CubeCache.

File: src/main/java/cubicchunks/server/PlayerCubeMap.java [hunk 3]
```diff
@@ -177,30 +178,15 @@ public void onConfigUpdate(CubicChunks.Config config) {
...
-					}
-					if (!column.isChunkTicked()) {
-						return column;
-					}
-					//is there any non-spectator player within 128 blocks distance?
-					if (watcher.hasPlayerMatchingInRange(128.0D, NOT_SPECTATOR)) {
+			@Override protected Chunk computeNext() {
+				while (chunkIt.hasNext()){
+					Column column = (Column)chunkIt.next();
+					if(column.shouldTick()){ // shouldTick is true when there Cubes with tickets the request to be ticked
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/PlayerCubeMap.java]: This should only return columns that PlayerCubeMap knows about. Cubes that are loaded by players. Other cubes should not be ticked.

File: src/main/java/cubicchunks/server/PlayerCubeMapEntry.java [hunk 4]
```diff
@@ -128,11 +135,17 @@ public boolean providePlayerCube(boolean canGenerate) {
 		int cubeZ = getZ(cubeAddress);
 
 		playerCubeMap.getWorld().getProfiler().startSection("getCube");
+		if(cube != null){
+			this.cube.getTickets().remove(this); // just in case there is a cube already loaded
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/PlayerCubeMapEntry.java]: Why does it even allow adding the same ticket twice?

File: src/main/java/cubicchunks/server/ServerCubeCache.java [hunk 5]
```diff
@@ -586,26 +489,51 @@ public void flush() {
...
 	}
 
-	public void unloadCube(Cube cube) {
-		// don't unload cubes near the spawn
-		if (cubeIsNearSpawn(cube)) {
-			return;
+	void takeOutGarbage(){
+
+		Iterator<Cube> cubeIt = cubeMap.iterator();
+		while(cubeIt.hasNext()) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/ServerCubeCache.java]: Unloading all cubes that can be unloaded every tick is a really bad idea for performance. Unloading can be quite slow. That's why it's queued.

File: src/main/java/cubicchunks/util/ticket/TicketList.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		}else if(!one.shouldTick() && two.shouldTick()){
+			return 1;
+		}
+		return 0;
+	};
+
+	private LinkedList<ITicket> list = new LinkedList<>();
+
+	public TicketList(){
+		super();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/util/ticket/TicketList.java]: That's not needed

File: src/main/java/cubicchunks/util/ticket/TicketList.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	/**
+	 * Add a ticket to this list
+	 *
+	 * @param ticket the ticket to add
+	 */
+	public void add(ITicket ticket){
+		list.add(ticket);
+		list.sort(ORDER);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/util/ticket/TicketList.java]: Instead of sorting it would be more efficient to iterate over all of them here and update shouldTick flag.

File: src/main/java/cubicchunks/world/ICubicWorld.java [hunk 8]
```diff
@@ -139,7 +140,7 @@ default boolean testForCubeAndNeighbor(Cube cube, @Nullable Predicate<Cube> test
 
 	int getEffectiveHeight(int blockX, int blockZ);
 
-	//this is a hack
+	//this is a hack TODO: remove!
 	void setGeneratingWorld(boolean generating);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/ICubicWorld.java]: This hack was my attempt to force light updates when generating cubes. It's not used anymore.

File: src/main/java/cubicchunks/world/column/CubeMap.java [hunk 9]
```diff
@@ -23,77 +23,135 @@
...
 	}
 
-	ExtendedBlockStorage[] getStorageArrays() {
-		//TODO: automatically update set if there is a Cube with a new ExtendedBlockStorage
-		return set.getArray(); 
+	/**
+	 * @return An array of EBS's form Cubes that need ticking... kind of a hack but vanilla needs it
+	 */
+	ExtendedBlockStorage[] getStoragesToTick() {
+		if(!isToTickValid()) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/column/CubeMap.java]: no no no no

Vanilla assumes this is instant because it just returns the internal array. Doing that here isn't going to be fast.

File: src/main/java/cubicchunks/world/cube/Cube.java [hunk 10]
```diff
@@ -571,6 +568,14 @@ public void onUnload() {
 		// tell the world to forget about entities
 		this.world.unloadEntities(this.entities.getEntities());
 
+		for(Entity entity : this.entities.getEntities()){
+			//CHECKED: 1.10.2-12.18.1.2092
+			entity.addedToChunk = false; // World tries to remove entities from Cubes
+										 // if (addedToCube || Column is loaded)
+										 // so we need to set addedToChunk to false as a hack!
+										 // else World would reload this Cube!
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/cube/Cube.java]: if you want to line up comments, use spaces. Or don't do it at all.

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorldServer.java [hunk 11]
```diff
@@ -79,47 +79,13 @@
...
-		int generated = 0;
-
-		for (int cubeX = spawnCubeX - spawnDistance; cubeX <= spawnCubeX + spawnDistance; cubeX++) {
-			for (int cubeZ = spawnCubeZ - spawnDistance; cubeZ <= spawnCubeZ + spawnDistance; cubeZ++) {
-				for (int cubeY = spawnCubeY + spawnDistance; cubeY >= spawnCubeY - spawnDistance; cubeY--) {
-					serverCubeCache.getCube(cubeX, cubeY, cubeZ, IProviderExtras.Requirement.LIGHT);
-					generated++;
-					if (System.currentTimeMillis() >= lastTime + progressReportInterval) {
-						lastTime = System.currentTimeMillis();
-						CubicChunks.LOGGER.info("Preparing spawn area: {}%", generated*100/totalToGenerate);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorldServer.java]: Didn't I tell you to add spawn chunks back?

File: src/main/java/cubicchunks/world/ICubicWorld.java (hunk 8 above)
Comment (xcube16) [inline src/main/java/cubicchunks/world/ICubicWorld.java]: I wanted that to be removed in another PR, so I just added the TODO

File: src/main/java/cubicchunks/world/cube/Cube.java [hunk 12]
```diff
@@ -645,13 +654,6 @@ public boolean isInitialLightingDone() {
 		return isInitialLightingDone;
 	}
 
-	/**
-	 * Lights up all blocks that will definitely be lit. It's faster than using world.checkLightFor on everything.
-	 */
-	public void initSkyLight() {
-		((ICubicWorldServer)this.getWorld()).getFirstLightProcessor().initializeSkylight(this);
```
Comment (xcube16) [inline src/main/java/cubicchunks/world/cube/Cube.java]: initializeSkylight() is called directly in ServerCubeCache now

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java [hunk 13]
```diff
@@ -67,10 +67,12 @@ private boolean canUpdateEntity(World _this, int startBlockX, int oldStartBlockY
...
 		}
+
+		BlockPos entityPos = new BlockPos(updateEntity_entityPosX, updateEntity_entityPosY, updateEntity_entityPosZ);
+
 		return this.isRemote() ||
-				this.getCubeFromBlockCoords(
-						new BlockPos(updateEntity_entityPosX, updateEntity_entityPosY, updateEntity_entityPosZ))
-						.getTickets().shouldTick();
+			  !this.isValid(entityPos) || // if an entity is below the world we want it to keep ticking
+			   this.getCubeFromBlockCoords(entityPos).getTickets().shouldTick();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java]: Without any buffer radius bigger entities will still load chunks because of physics calculations

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java (hunk 1 above)
Comment (xcube16) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java]: Ya, it was a bit of a mess...
If you have an entity in a world at x, y, z is'nt the Cube at x, y, z loaded? Or would the entity be gone when the server restarts?
As for the 32 block thing... If a ticket says things need to be ticked, they really NEED to be ticked.

File: src/main/java/cubicchunks/server/ChunkGc.java (hunk 2 above)
Comment (xcube16) [inline src/main/java/cubicchunks/server/ChunkGc.java]: Moved it back to ChunkGC :)

**Comment by Barteks2x** [comment]
Stupid github thinks my comment is outdated when it isn't... (the one above)

File: src/main/java/cubicchunks/server/PlayerCubeMap.java (hunk 3 above)
Comment (xcube16) [inline src/main/java/cubicchunks/server/PlayerCubeMap.java]: PlayerCubeMap.getChunkIterator() is just being used like an event hook now...
Tickets now control when Cubes should be ticked, (PlayerCubeMap uses tickets now so its ok)

File: src/main/java/cubicchunks/server/PlayerCubeMapEntry.java (hunk 4 above)
Comment (xcube16) [inline src/main/java/cubicchunks/server/PlayerCubeMapEntry.java]: Ya, that was a bad idea... fixing it

File: src/main/java/cubicchunks/util/ticket/TicketList.java (hunk 7 above)
Comment (xcube16) [inline src/main/java/cubicchunks/util/ticket/TicketList.java]: Good idea!

File: src/main/java/cubicchunks/world/column/CubeMap.java (hunk 9 above)
Comment (xcube16) [inline src/main/java/cubicchunks/world/column/CubeMap.java]: getStoragesToTick() basicly gets a list of Cubes that need ticking (That is what vanilla uses it for)...
Should I make it so that the array is updated when a cube gets a ticket that requests ticking?
(right now it does it dynamicly)

File: src/main/java/cubicchunks/world/column/CubeMap.java [hunk 14]
```diff
@@ -23,77 +23,135 @@
...
+		}else{
+			return Collections.emptyList();
 		}
-		return this.cubeMap.subMap(startY, true, endY, true).values();
 	}
 
 	private boolean contains(int cubeY) {
-		return this.map.containsKey(cubeY);
+		int index = binarySearch(cubeY);
+		return index < cubes.size() && cubes.get(index).getY() == cubeY;
```
Comment (xcube16) [inline src/main/java/cubicchunks/world/column/CubeMap.java]: If someone wants to get a Cube they should use the main cube map... (And that is what Column does).
So this is not a hot path

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java (hunk 13 above)
Comment (xcube16) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java]: What is wrong with that? An entity loads only what it needs... aka no forced radius requirement

**Comment by Barteks2x** [comment]
@xcube16 the reason you had issues with entity ticking was that vertical view distance is too small, even if GUI says it's 8.

**Comment by xcube16** [comment]
*crash*
um... one sec, I need to fix something

**Comment by xcube16** [comment]
> @xcube16 the reason you had issues with entity ticking was that vertical view distance is too small, even if GUI says it's 8.

It was only half the problem... The issue that this fixes is that I could see frozen entities (this was bad)
Now, you see Cube, you see live entities. :)

**Comment by Barteks2x** [comment]
Entities should not be ticked when they are close to unloaded chunks. That's the way it works in vanilla, and we should do it that way. 

Forced chunks are exception because they are added by forge and they are those from chunkloaders.

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java (hunk 13 above)
Comment (Barteks2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java]: Add that 32 block radius back. This really is needed. And stop with that "the world is infinite" thing. It's not it's running on real machine with finite amount of memory and slow disks.

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java (hunk 1 above)
Comment (Barteks2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java]: Entities should not be ticked unless there are chunks around it loaded. That's how it works, and don't change it. Forge forced chunks are exception to make chunkloaders work.

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorldServer.java (hunk 11 above)
Comment (Barteks2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorldServer.java]: It's removed in this PR now, it should be there.

File: src/main/java/cubicchunks/server/SpawnCubes.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
+package cubicchunks.server;
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/SpawnCubes.java]: license header missing

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java [hunk 16]
```diff
@@ -67,18 +67,12 @@ private boolean canUpdateEntity(World _this, int startBlockX, int oldStartBlockY
...
+
 		BlockPos entityPos = new BlockPos(updateEntity_entityPosX, updateEntity_entityPosY, updateEntity_entityPosZ);
-		if (!isValid(entityPos)) {
-			return true;
-		}
-		return isAreaLoaded(startBlockX, startBlockY, startBlockZ, endBlockX, endBlockY, endBlockZ, allowEmpty);
+
+		return this.isRemote() ||
+			  !this.isValid(entityPos) || // if an entity is below the world we want it to keep ticking
+			   this.getCubeFromBlockCoords(entityPos).getTickets().shouldTick();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_Tick.java]: Whatever way you do it, entities near the edge of loaded area shouldn't be updated.

**Comment by Barteks2x** [comment]
Merging it without the 32 block radius with loaded chunks for entity ticking, but if it ever causes issues - the check is going back.
